aten/src/ATen/native/cuda/LegacyThrustHelpers.cu

```diff
@@ -30,4 +32,66 @@ void index_put_with_sort_kernel_thrust_helper(Tensor &linearIndex, Tensor &orig_
   thrust::sort_by_key(policy, sorted_data, sorted_data + num_indices, orig_data, LTOp<int64_t>());
 }
 
+template<typename index_t>
+int embedding_renorm_cuda_unique_copy(Tensor &indices_contig, Tensor &unique_indices) {
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  using device_ptr = thrust::device_ptr<index_t>;
+
+  auto num_indices = indices_contig.numel();
+  auto indices_data = device_ptr(indices_contig.data_ptr<index_t>());
+  auto unique_data = device_ptr(unique_indices.data_ptr<index_t>());
+  auto end = thrust::unique_copy(policy, indices_data, indices_data + num_indices, unique_data);
+  auto num_unique_indices = static_cast<int>(end - unique_data);
+  return num_unique_indices;
+}
+
+template
+int embedding_renorm_cuda_unique_copy<int>(Tensor &indices_contig, Tensor &unique_indices);
+template
+int embedding_renorm_cuda_unique_copy<int64_t>(Tensor &indices_contig, Tensor &unique_indices);
+
+template<typename index_t>
+void embedding_dense_backward_cuda_scan(Tensor &sorted_indices, Tensor &count) {
+  using device_ptr = thrust::device_ptr<index_t>;
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  auto num_indices = count.numel();
+
+  // Compute an increasing sequence per unique item in sortedIndices:
+  // sorted: 2 5 5 5 7 7 8 9 9
+  //  count: 1 1 2 3 1 2 1 1 2
+  auto sorted_data = device_ptr(sorted_indices.data_ptr<index_t>());
+  auto count_data = device_ptr(count.data_ptr<index_t>());
+  thrust::inclusive_scan_by_key(
+    policy,
+    sorted_data,
+    sorted_data + num_indices,
+    thrust::make_constant_iterator(1),
+    count_data
+  );
+
+  // Take the maximum of each count per unique key in reverse:
+  // sorted: 2 5 5 5 7 7 8 9 9
+  //  count: 1 3 3 3 2 2 1 2 2
+  thrust::inclusive_scan_by_key(
```

Comment:
I looked up implementation, and also looks like this is accessing cached device properties, so ideally should be moved to cub also, but unfortunately cub doesn't provide ready-made inverse iterators.

Reply:
Same here, in next PR :)